zerver/tests/test_external.py

```diff
@@ -118,6 +129,29 @@ def test_hit_ratelimits(self) -> None:
 
             self.assert_json_success(result)
 
+    def test_hit_ratelimits_as_ip(self) -> None:
+        RateLimitedIPAddr("127.0.0.1").clear_history()
+
+        start_time = time.time()
+        for i in range(6):
+            with mock.patch('time.time', return_value=(start_time + i * 0.1)):
+                result = self.send_unauthed_api_request()
+
+        self.assertEqual(result.status_code, 429)
+        json = result.json()
+        self.assertEqual(json.get("result"), "error")
+        self.assertIn("API usage exceeded rate limit", json.get("msg"))
+        self.assertEqual(json.get('retry-after'), 0.5)
+        self.assertTrue('Retry-After' in result)
+        self.assertEqual(result['Retry-After'], '0.5')
+
+        # We actually wait a second here, rather than force-clearing our history,
+        # to make sure the rate-limiting code automatically forgives a user
+        # after some time has passed.
```

Comment:
We should replace "actually wait" with "simulate waiting a second" ... since we're not actually doing a 1-second sleep.

Reply:
Fixed in the 3rd, refactor commit.